src/Microsoft.Health.Dicom.Api/Features/Partition/DataPartitionFeatureValidatorService.cs

```diff
@@ -0,0 +1,60 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Dicom.Core.Configs;
+using Microsoft.Health.Dicom.Core.Exceptions;
+using Microsoft.Health.Dicom.Core.Features.Partition;
+
+namespace Microsoft.Health.Dicom.Api.Features.Partition
+{
+    /// <summary>
+    /// This hosted service performs the check at startup to ensure user cannot disable DataPartitions feature flag
+    /// if they already created partitioned data other than the default partition.
+    /// If a user created partitioned data other than the default partition, then an exception is thrown to block startup.
+    /// We will allow users to enable DataPartitions feature even if they have already created data,
+    /// it will be accessible by specifying Microsoft.Default partition name in the request.
+    /// </summary>
+    public class DataPartitionFeatureValidatorService : IHostedService
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly bool _isPartitionEnabled;
+
+        public DataPartitionFeatureValidatorService(
+            IServiceProvider serviceProvider,
+            IOptions<FeatureConfiguration> featureConfiguration)
+        {
+            _serviceProvider = serviceProvider;
+            EnsureArg.IsNotNull(featureConfiguration, nameof(featureConfiguration));
+            _isPartitionEnabled = featureConfiguration.Value.EnableDataPartitions;
+        }
+
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            if (!_isPartitionEnabled)
+            {
+                using (var scope = _serviceProvider.CreateScope())
```

Comment:
Do we need the new service scope for `IPartitionService`?

Reply:
Yes we need because we cant inject a scoped service in a hosted services. The service needs to be singleton.